src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,985 @@
+                xTaskNotifyGive(wifiTaskHandle_);
+
+                // If we are managing the WiFi and MDNS systems we need to
+                // trigger the reconnection process at this point.
+                if(manageWiFi_)
+                {
+                    LOG(INFO, "[WiFi] Attempting to reconnect to SSID: %s.",
+                        ssid_);
+                    esp_wifi_connect();
+                }
+            }
+            break;
+    }
+}
+
+// If the Esp32WiFiManager is setup to manage the WiFi system, the following
+// steps are executed:
+// 1) Start the TCP/IP adapter.
+// 2) Hook into the ESP event loop so we receive WiFi events.
+// 3) Initialize the WiFi system.
+// 4) Set the WiFi mode to STATION (WIFI_STA)
+// 5) Configure the WiFi system to store parameters only in memory to avoid
+// potential corruption of entries in NVS.
+// 6) Configure the WiFi system for SSID/PW.
+// 7) Set the hostname based on the generated hostname.
+// 8) Connect to WiFi and wait for DHCP IP assignment.
+// 9) Verify that we connected and received a DHCP IP address, if not log a
+// FATAL message and give up.
+// 10) Initialize the mDNS system.
+// 11) Set the mDNS hostname based on the generated hostname.
+// 12) Set the default mDNS instance name based on the generated hostname.
+void Esp32WiFiManager::start_wifi_system()
+{
+    // Create the event group used for tracking connected/disconnected status.
+    // This is used internally regardless of if we manage the rest of the WiFi
+    // or mDNS systems.
+    wifiStatusEventGroup_ = xEventGroupCreate();
+
+    // If we do not need to manage the WiFi and mDNS systems exit early.
+    if(!manageWiFi_)
+    {
+        return;
+    }
+
+    // Initialize the TCP/IP adapter stack.
+    LOG(INFO, "[WiFi] Starting TCP/IP stack");
+    tcpip_adapter_init();
+
+    // Install event loop handler.
+    ESP_ERROR_CHECK(esp_event_loop_init(wifi_event_handler, this));
+
+    // Start the WiFi adapter.
+    wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
+    LOG(INFO, "[WiFi] Initializing WiFi stack");
+    ESP_ERROR_CHECK(esp_wifi_init(&cfg));
+
+    // Set the WiFi mode to STATION.
+    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+
+    // This disables storage of SSID details in NVS which has been shown to be
+    // problematic at times for the ESP32, it is safer to always pass fresh
+    // config and have the ESP32 resolve the details at runtime rather than
+    // use a cached set from NVS.
+    esp_wifi_set_storage(WIFI_STORAGE_RAM);
+
+    // Configure the SSID details for the station based on the SSID and
+    // password provided to the Esp32WiFiManager constructor.
+    wifi_config_t conf;
+    memset(&conf, 0, sizeof(wifi_config_t));
+    strcpy(reinterpret_cast<char*>(conf.sta.ssid), ssid_);
+    // if we have a password check if it is a PSK or regular password
+    if(password_)
+    {
+        if (strlen(password_) == 64) // it's the PSK and not a password.
```

Comment:
What is the difference between the two branches? They both seem to copy until the first null.
Seems like this could be replaced with a single strncpy

Reply:
you're right, it doesn't make sense as written. The code was copied as-is from the arduino code and doesn't really make sense there either. will consolidate to just the else case instead.